### PR TITLE
hotplug_mem_reserve: add new case to test reserve option

### PR DESCRIPTION
**Pull request by huangyum**
Signed-off-by: Yumei Huang <yuhuang@redhat.com>
ID: 2012092
Depends on https://github.com/avocado-framework/avocado-vt/pull/3285.

### CODE REVIEW

**Comment by huangyum** [comment]
(1/1) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.hotplug_memory.after.guest_reboot.unplug.pluged_memory.backend_ram.no_policy.two.op_on: PASS (183.14 s)
@yanan-fu @MiriamDeng Would you pls help review, thanks.

**Comment by MiriamDeng** [comment]
LGTM
git am 3104.patch
Applying: hotplug_mem: add option 'reserve' test
JOB LOG    : /root/avocado/job-results/job-2021-11-29T20.28-b34ecae/job.log
 (1/1) Host_RHEL.m8.u6.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.hotplug_memory.after.guest_reboot.unplug.pluged_memory.backend_ram.no_policy.two.op_on: PASS (276.80 s)

**Comment by yanan-fu** [comment]
No concern with this patch, just waiting for the avocado-vt patch now, thanks!

**Comment by huangyum** [comment]
@yanan-fu  I need update a step after some discussion with dev. Will let you know once it's ready. Thanks.

**Comment by huangyum** [comment]
(1/4) Host_RHEL.m8.u6.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.6.0.x86_64.io-github-autotest-qemu.hotplug_mem_reserve.reserve_on: PASS (74.18 s)
 (2/4) Host_RHEL.m8.u6.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.6.0.x86_64.io-github-autotest-qemu.hotplug_mem_reserve.reserve_off: PASS (75.14 s)
 (3/4) Host_RHEL.m8.u6.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.hotplug_mem_reserve.reserve_on: PASS (212.13 s)
 (4/4) Host_RHEL.m8.u6.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.hotplug_mem_reserve.reserve_off: PASS (132.10 s)

@MiriamDeng @yanan-fu Just updated the steps, would you pls help review again, thanks.

**Comment by huangyum** [comment]
@MiriamDeng Would you pls help test on ppc again? The case steps change a lot. Thanks.

**Comment by MiriamDeng** [comment]
LGTM 
 (1/1) Host_RHEL.m8.u6.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.hotplug_memory.after.guest_reboot.unplug.pluged_memory.backend_ram.no_policy.two.op_on: PASS (252.33 s)

**Comment by MiriamDeng** [comment]
Test result on ppc,
(1/1) Host_RHEL.m8.u6.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.hotplug_memory.after.guest_reboot.unplug.pluged_memory.backend_ram.no_policy.two.op_on: PASS (252.33 s)

 (1/2) Host_RHEL.m8.u6.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.6.0.ppc64le.io-github-autotest-qemu.hotplug_mem_reserve.reserve_on: ERROR: VM creation command failed:    "MALLOC_PERTURB_=1  /usr/libexec/qemu-kvm -S  -name 'avocado-vt-vm1'  -sandbox on  -machine pseries  -nodefaults -device VGA,bus=pci.0,addr=0x2 -m 4096,maxmem=32G,slots=4 -object memory-backend-memfd,hugetlb=yes,reserve=yes,... (6.30 s)
 (2/2) Host_RHEL.m8.u6.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.6.0.ppc64le.io-github-autotest-qemu.hotplug_mem_reserve.reserve_off: ERROR: VM creation command failed:    "MALLOC_PERTURB_=1  /usr/libexec/qemu-kvm -S  -name 'avocado-vt-vm1'  -sandbox on  -machine pseries  -nodefaults -device VGA,bus=pci.0,addr=0x2 -m 4096,maxmem=32G,slots=4 -object memory-backend-memfd,hugetlb=yes,reserve=no,s... (7.31 s)
R

**Comment by huangyum** [comment]
@MiriamDeng Would you pls have a try with both avocado-framework/avocado-vt#3285 and avocado-framework/avocado-vt#3306 ? Thanks!

File: qemu/tests/hotplug_mem_reserve.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+    vm = env.get_vm(params["main_vm"])
+    vm.wait_for_login()
+    mem_name = params["target_mems"]
+    hotplug_test = MemoryHotplugTest(test, params, env)
+    hotplug_test.hotplug_memory(vm, mem_name)
+    hotplug_test.check_memory(vm)
+    time.sleep(30)
```
Comment (yanan-fu) [inline qemu/tests/hotplug_mem_reserve.py:27]: I wonder if there is any effective method to check it is ready or not ?

File: qemu/tests/cfg/hotplug_mem_reserve.cfg [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    size_mem_mem0 = ${mem}M
+    numa_memdev_node0 = mem-mem0
+    backend_mem = memory-backend-memfd
+    target_mems = plug1
+    size_mem_plug1 = 1G
+    hugetlb_mem = yes
+    setup_hugepages = yes
+    extra_hp_size = ${size_mem_plug1}
+    ppc64,ppc64le:
+        threshold = 0.15
```
Comment (yanan-fu) [inline qemu/tests/cfg/hotplug_mem_reserve.cfg:21]: No calling for `threshold`

Comment (huangyum) [reply qemu/tests/cfg/hotplug_mem_reserve.cfg:21]: Yes, it's used in vt.

**Comment by MiriamDeng** [comment]
there's still something wrong on ppc
 (1/2) Host_RHEL.m8.u6.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.6.0.ppc64le.io-github-autotest-qemu.hotplug_mem_reserve.reserve_on: ERROR: QMP command 'object-add' failed    (arguments: {'qom-type': 'memory-backend-memfd', 'id': 'mem-plug1', 'hugetlb': True, 'reserve': True, 'size': 1073741824},    error message: {'class': 'GenericError', 'desc': 'unable to map backing store for guest RAM: C... (108.56 s)
 (2/2) Host_RHEL.m8.u6.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.6.0.ppc64le.io-github-autotest-qemu.hotplug_mem_reserve.reserve_off: PASS (139.45 s)

File: qemu/tests/hotplug_mem_reserve.py (hunk 1 above)
Comment (huangyum) [reply qemu/tests/hotplug_mem_reserve.py:27]: Maybe use wait_for to check HugePages_Rsvd every 5 seconds untils it's stable?

**Comment by MiriamDeng** [comment]
Test passed on ppc
 (1/2) Host_RHEL.m8.u6.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.6.0.ppc64le.io-github-autotest-qemu.hotplug_mem_reserve.reserve_on: PASS (113.87 s)
 (2/2) Host_RHEL.m8.u6.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.6.0.ppc64le.io-github-autotest-qemu.hotplug_mem_reserve.reserve_off: PASS (140.09 s)

File: qemu/tests/hotplug_mem_reserve.py (hunk 1 above)
Comment (yanan-fu) [reply qemu/tests/hotplug_mem_reserve.py:27]: Agree with that

File: qemu/tests/hotplug_mem_reserve.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+    vm = env.get_vm(params["main_vm"])
+    vm.wait_for_login()
+    mem_name = params["target_mems"]
+    hotplug_test = MemoryHotplugTest(test, params, env)
+    hotplug_test.hotplug_memory(vm, mem_name)
+    hotplug_test.check_memory(vm)
```
Comment (yanan-fu) [inline qemu/tests/hotplug_mem_reserve.py:26]: From the manual test case, memdev options need be checked to,  but i did not see this step, is this intentional ?

Comment (huangyum) [reply qemu/tests/hotplug_mem_reserve.py:26]: It's checked inside check_memory().

File: qemu/tests/cfg/hotplug_mem_reserve.cfg (hunk 2 above)
Comment (yanan-fu) [reply qemu/tests/cfg/hotplug_mem_reserve.cfg:21]: Got it

**Comment by huangyum** [comment]
(1/2) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.hotplug_mem_reserve.reserve_on: PASS (43.94 s)
 (2/2) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.hotplug_mem_reserve.reserve_off: PASS (41.12 s)
@yanan-fu Pls help review again, thanks.

File: qemu/tests/hotplug_mem_reserve.py (hunk 3 above)
Comment (yanan-fu) [reply qemu/tests/hotplug_mem_reserve.py:38]: Wow, so deap, got it now, thanks!

**Comment by yanan-fu** [comment]
To test hotplug mem with `reserve=***`,   how about the default boot mem ?  Is it must to use memory-backend-memfd and hugetlb=on ?

**Comment by huangyum** [comment]
> To test hotplug mem with `reserve=***`, how about the default boot mem ? Is it must to use memory-backend-memfd and hugetlb=on ?

Yes, it's recommended to test with these configuration.

**Comment by yanan-fu** [comment]
No more comments, waiting for avocado-vt patch, thanks!

**Comment by huangyum** [comment]
Test results on x86: 
(1/2) Host_RHEL.m8.u6.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.hotplug_mem_reserve.reserve_on.q35: PASS (230.97 s)
 (2/2) Host_RHEL.m8.u6.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.hotplug_mem_reserve.reserve_off.q35: PASS (259.31 s)

Test results on ppc:
 (1/2) Host_RHEL.m8.u6.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.6.0.ppc64le.io-github-autotest-qemu.hotplug_mem_reserve.reserve_on: PASS (110.22 s)
 (2/2) Host_RHEL.m8.u6.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.6.0.ppc64le.io-github-autotest-qemu.hotplug_mem_reserve.reserve_off: PASS (87.97 s)

**Review by YongxueHong — APPROVED** [review]

**Pull request merged**